src/Storages/MergeTree/DataPartsExchange.cpp

```diff
@@ -141,6 +141,16 @@ void Service::processQuery(const HTMLForm & params, ReadBuffer & /*body*/, Write
 
         CurrentMetrics::Increment metric_increment{CurrentMetrics::ReplicatedSend};
 
+        {
+            auto disk = part->volume->getDisk();
+            UInt64 revision = parse<UInt64>(params.get("disk_revision", "0"));
+            if (revision)
+                disk->syncRevision(revision);
+            revision = disk->getRevision();
```

Comment:
We have no collision because now every revision has hostname as additional part. Counter in ZooKeeper would have a very high cost, because it required request to ZooKeeper for every S3 operation. Second reason is that revision number is a part of DiskS3 realization and should work for single ClickHouse without ZooKeeper. Now DiskS3 knows nothing about ZooKeeper and it's right in my opinion.

Reply:
But it's just doesn't work. We have a lot of places in code where revision_counter incremented, a lot of tables leave on the same disk. So what we are trying to achieve? While we are sending it over network it can change 100500 times.

Reply:
That does not matter. Purpose of this sync is:
1. when replica 1 creates file and makes some operations
2. after that replica 2 fetches this file
3. after that replica 2 makes some operations with this file
then revisions from step 3 are always after revisions from step 1.
In other case may be situation:
- replica 1 creates file `file1_tmp` with object `r42_1_file1_tmp` (42 is a revision number here)
- replica 1 renames it in `file1` with operation in log 43
- replica 2 fetches this file
- replica 2 renames `file` to `file1_detached` with operation 21 (less than 42)
- replica 2 makes backup
- new cluster restoring file `file1_tmp` from object `r42_1_file1_tmp`
- new cluster is restoring operations on revision order, trying to rename `file1` to `file1_detached`, and this fails, because local file `file1` is not exists yet, it will be only after revision 43.